libres/lib/analysis/ies/ies.cpp

```diff
@@ -229,162 +218,142 @@ void ies::updateA(
-    matrix_type *AAi = matrix_alloc(ens_size, ens_size);
-    matrix_subtract_row_mean(Ai);
-    matrix_type *VT = matrix_alloc(state_size, ens_size);
+    Eigen::MatrixXd Ai = A;
+    matrix_subtract_row_mean(&Ai);
 
-    auto svd = Ai->bdcSvd(Eigen::ComputeThinV);
-    *VT = svd.matrixV().transpose();
+    auto svd = Ai.bdcSvd(Eigen::ComputeThinV);
+    Eigen::MatrixXd VT = svd.matrixV().transpose();
 
-    *AAi = VT->transpose() * *VT;
+    Eigen::MatrixXd AAi = VT.transpose() * VT;
 
-    matrix_inplace_matmul(Y, AAi);
-    matrix_free(Ai);
-    matrix_free(AAi);
-    matrix_free(VT);
+    Y *= AAi;
 }
 
 /*
 * COMPUTE  Omega= I + W (I-11'/sqrt(ens_size))    from Eq. (36).                                   (Line 6)
 *  When solving the system S = Y inv(Omega) we write
 *     Omega^T S^T = Y^T
 */
-void ies::linalg_solve_S(const matrix_type *W0, const matrix_type *Y,
-                         matrix_type *S) {
-    int ens_size = matrix_get_columns(W0);
-    int nrobs = matrix_get_rows(S);
+void ies::linalg_solve_S(const Eigen::MatrixXd &W0, const Eigen::MatrixXd &Y,
+                         Eigen::MatrixXd &S) {
+    int ens_size = W0.cols();
+    int nrobs = S.rows();
     double nsc = 1.0 / sqrt(ens_size - 1.0);
 
-    matrix_type *YT =
-        matrix_alloc(ens_size, nrobs); // Y^T used in linear solver
-    matrix_type *ST =
-        matrix_alloc(ens_size, nrobs); // current S^T used in linear solver
-    matrix_type *Omega =
-        matrix_alloc(ens_size, ens_size); // current S^T used in linear solver
+    Eigen::MatrixXd YT =
+        Eigen::MatrixXd::Zero(ens_size, nrobs); // Y^T used in linear solver
+    Eigen::MatrixXd ST =
+        Eigen::MatrixXd::Zero(ens_size, nrobs); // current S^T used in linear solver
 
     /*  Here we compute the W (I-11'/N) / sqrt(N-1)  and transpose it).*/
-    matrix_assign(
-        Omega,
-        W0); // Omega=data->W (from previous iteration used to solve for S)
-    matrix_subtract_row_mean(Omega);   // Omega=Omega*(I-(1/N)*11')
-    matrix_scale(Omega, nsc);          // Omega/sqrt(N-1)
-    matrix_inplace_transpose(Omega);   // Omega=transpose(Omega)
+    // Omega=data->W (from previous iteration used to solve for S)
+    Eigen::MatrixXd Omega = W0; // current S^T used in linear solver
+    matrix_subtract_row_mean(&Omega);   // Omega=Omega*(I-(1/N)*11')
+    Omega *= nsc;          // Omega/sqrt(N-1)
+    Omega.transposeInPlace();   // Omega=transpose(Omega)
     for (int i = 0; i < ens_size; i++) // Omega=Omega+I
-        matrix_iadd(Omega, i, i, 1.0);
+        Omega(i, i) += 1.0;
 
-    matrix_transpose(Y, YT); // RHS stored in YT
+    YT = Y.transpose(); // RHS stored in YT
 
     /* Solve system                                                                                (Line 7)   */
-    Eigen::MatrixXd tmp = Omega->fullPivLu().solve(*YT);
-    *YT = tmp;
-
-    matrix_transpose(YT, S); // Copy solution to S
+    Eigen::MatrixXd tmp = Omega.fullPivLu().solve(YT);
```

Comment:
Do we need to go via `tmp` or could we just do `Eigen::MatrixXd YT = Omega.fullPivLu().solve(Y.transpose())`?